### DISCUSSION TITLE
InvalidatePlot() seems to freeze UI briefly

### DISCUSSION THREAD

**Discussion by dknippel** [oxyplot/oxyplot#1910] [Q&A] [answered]
Hello,

I have a `PlotModel` with a single `ExtrapolationLineSeries` series with less than 300 DataPoints, which I'm guessing is not very many.  I first call **_InvalidatePlot(false)_**, then add a new `DataPoint`, then call **_InvalidatePlot(true)_**.  I do this for each said `DataPoint`.  Please note that said procedure is performed on a thread NOT sired by the UI.

This seems to cause approx 5 second freeze of the UI.

Am I using the `InvalidatePlot()` method incorrectly?  Is this perhaps unique to the  `ExtrapolationLineSeries` series?  I did not observe said behavior when using the regular `LineSeries` series.  Does @VisualMelon perhaps have any insight on this unique behavior?

Thanks,

**Comment by VisualMelon** [comment]
Which UI provider are you using (WPF, Winforms, Avalonia, etc.)? Are you able to provide your code?

I don't see any dramatic issue like you describe with a simple test in winforms and over a thousand points. `ExtrapolationLineSeries` does more work than `LineSeries`, and presently re-renders the whole plot for each interval and region between intervals, as I recall. You're probably also using a non-solid line for extrapolated regions, which tend to be more expensive to render than solid lines. If you're able to attach a profiler/debugger to your code, it will be easier to see what it's doing when seized (e.g. is the UI rendering itself, or is `ExtrapolationLineSeries` working too hard, etc.).

Note that if you are adding many data-points at once, you only need to call `InvalidePlot(true)` once you've added all of them (but in any case, this _shouldn't_ be a problem). There is no need to call `InvalidatePlot(false)` before making any changes. It may be that the provision for the platform you are using has a problem.

**Comment by dknippel** [comment, marked as answer]
Hi @VisualMelon , thank you for your reply.

I am using the WPF UI provider.  I was able to address the "freezing" by capturing all the `DataPoint` first and then using `Points.AddRange()`, rather than adding them individually.

I've tried only using `InvalidePlot(true)` , but the control becomes unstable and often throws the 'sequence contains no points' / 'collection was modified' exceptions if I don't call `InvalidePlot(false)` prior to adding/removing `DataPoint`,  'PointAnnotation',  `DataRange`,   etc.  Question in general:  What is the intent/function of calling `InvalidePlot(false)` ?

Also, when setting up Intervals what are the rules for the `DataRange` objects?  Are the `Minimum` and `Maximum` members inclusive/exclusive?  Can said intervals overlap?

Thanks again for all of your tribal knowledge insight; especially WRT the `ExptrapolationLineSeries` class (read:  not much documentation anywhere :-( )

**Reply by VisualMelon** [reply]
`InvalidatePlot(false)` must be called after you make 'cosmetic'/'view' changes; `InvalidePlot(true)` must be called if you change any data (`InvalidatePlot(true)` does everything that `InvalidePlot(false)` does as well, so never a need to call both).

If you are getting those errors, it suggests you are updating the model from a different thread, or otherwise at the same time as it is updating or rendering. It is important that you take the model's `SyncRoot` when changing anything, and invalid the plot while holding the monitor to avoid these issues:

    // take the monitor
    lock (myPlotModel.SyncRoot)
    {
        // perform any updates
        mySeries.DataPoints.Add(new DataPoint(0, 1));

        // finally, invalidate the plot
        myPlotModel.InvalidatePlot(updateData);
    }

Re. intervals, they actually only affect the rendering, and I suspect they are effectively inclusive, but I don't think it draws a real distinction here.

**Reply by dknippel** [reply]
OK, that's helpful.  Will apply your suggestions and see if that addresses the stability issue...
